pkg/plugin/validator.go

```diff
@@ -0,0 +1,171 @@
+	"github.com/google/go-github/v55/github"
+	"github.com/lestrrat-go/jwx/v2/jwk"
+)
+
+// Validator validates github issue against validation criteria.
+type Validator struct {
+	cfg        *PluginConfig
+	decodedPem *rsa.PrivateKey
+	client     *github.Client
+
+	// for testing only
+	fakeGithubIssue *github.Issue
+}
+
+// ExchangeResponse is the GitHub API response of requesting an access token
+// for the GitHub App installation with requested repositories and permissions.
+type ExchangeResponse struct {
+	AccessToken string `json:"token"`
+}
+
+// pluginGithubIssue contains the required attribute parsed from
+// the issue URL.
+type pluginGithubIssue struct {
+	Owner       string
+	RepoName    string
+	IssueNumber int
+}
+
+// NewValidator creates a validator.
+func NewValidator(cfg *PluginConfig) (*Validator, error) {
+	pk, err := readPrivateKey(cfg.GitHubAppPrivateKeyPEM)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read private key: %w", err)
+	}
+	v := Validator{
+		cfg:        cfg,
+		decodedPem: pk,
+		client:     github.NewClient(nil),
+	}
+	return &v, nil
+}
+
+// MatchIssue parses issue info from provided issueURL and validate if the issue is valid.
+func (v *Validator) MatchIssue(ctx context.Context, issueURL string, opts ...githubapp.ConfigOption) error {
+	info, err := parseIssueInfoFromURL(issueURL)
+	if err != nil {
+		return fmt.Errorf("failed to parse issueURL: %w", err)
+	}
+
+	t, err := v.getAccessToken(ctx, v.cfg, v.decodedPem, info.RepoName, opts)
+	if err != nil {
+		return fmt.Errorf("failed to get access token: %w", err)
+	}
+	v.client = v.client.WithAuthToken(t)
+
+	return v.validateIssue(ctx, info)
+}
+
+// validateIssue verifies if the issue exists and the issue is open.
+func (v *Validator) validateIssue(ctx context.Context, pi *pluginGithubIssue) error {
+	issue, err := v.getGithubIssue(ctx, pi)
+	if err != nil {
+		return fmt.Errorf("failed to get issue info: %w", err)
+	}
+	if s := issue.GetState(); s != "open" {
+		return fmt.Errorf("issue is in state: %s, please make sure to use an open issue", s)
+	}
+	return nil
+}
+
+// getGithubIssue gets the provided issue's info from github api.
+func (v *Validator) getGithubIssue(ctx context.Context, pi *pluginGithubIssue) (*github.Issue, error) {
+	if v.fakeGithubIssue != nil {
+		return v.fakeGithubIssue, nil
+	}
+	issue, _, err := v.client.Issues.Get(ctx, pi.Owner, pi.RepoName, pi.IssueNumber)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get issue: %w", err)
+	}
+	return issue, nil
+}
+
+// getAccessToken gets an access token with issue read permission to the repo
+// which contains the issue.
+func (v *Validator) getAccessToken(ctx context.Context, cfg *PluginConfig, pk *rsa.PrivateKey, repoName string, opts []githubapp.ConfigOption) (string, error) {
+	opts = append(opts, githubapp.WithJWTTokenCaching(1*time.Minute))
+
+	ghCfg := githubapp.NewConfig(cfg.GitHubAppID, cfg.GitHubAppInstallationID, pk, opts...)
+	githubApp := githubapp.New(ghCfg)
```

Comment:
Another option is to replace cfg *PluginConfig with githubapp, so that githubapp is created outside of validator and people can add whatever options to it instead of pass it via opts []githubapp.ConfigOption which seems strange.

Reply:
See other comment: https://github.com/abcxyz/jvs-plugin-github/pull/2#discussion_r1373839108